contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+
+### CRI API Changes
+Proposed CRI API changes:
+
+```golang
+// Runtime service defines the public APIs for remote container runtimes
+service RuntimeService {
+    // Version returns the runtime name, runtime version, and runtime API version.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+    …….
+    …….
+    //  RuntimeConfigInfo returns the configuration details of the runtime.
+    rpc RuntimeConfigInfo(RuntimeConfigInfoRequest) returns (RuntimeConfigInfoResponse) {}
+}
+// LinuxIDMapping represents a single user namespace mapping in Linux.
+message LinuxIDMapping {
+   // container_id is the starting id for the mapping inside the container.
+   uint32 container_id = 1;
+   // host_id is the starting id for the mapping on the host.
+   uint32 host_id = 2;
+   // size is the length of the mapping.
+   uint32 size = 3;
+}
+
+message UserNamespaceConfig {
+   // is_enabled, if true indicates that user-namespaces are supported and enabled in the container runtime
+   bool is_enabled = 1;
+   // uid_mappings is an array of user id mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group id mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+message RuntimeConfig {
+    NetworkConfig network_config = 1;
+    UserNamespaceConfig user_namespace_config = 2;
+}
+
+message UpdateRuntimeConfigRequest {
+    RuntimeConfig runtime_config = 1;
+}
+
+message UpdateRuntimeConfigResponse {}
+
+message RuntimeConfigInfoRequest {}
+
+message RuntimeConfigInfoResponse {
+    RuntimeConfig runtime_config = 1
+}
+
+...
+
+// NamespaceOption provides options for Linux namespaces.
+message NamespaceOption {
+	// Network namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode network = 1;
+	// PID namespace for this container/sandbox.
+	// Note: The CRI default is POD, but the v1.PodSpec default is CONTAINER.
+	// The kubelet's runtime manager will set this to CONTAINER explicitly for v1 pods.
+	// Namespaces currently set by the kubelet: POD, CONTAINER, NODE
+	NamespaceMode pid = 2;
+	// IPC namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped IPC in the Kubernetes API.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode ipc = 3;
+	// User namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped user namespace in the
+	// Kubernetes API. The container runtime has to run with user namespace enabled,
+	// and kubelet then forces the use of host user namespace.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceModeValue user = 4;
+}
+
+// NamespaceModeValue is the wrapper of NamespaceMode, to allow non-POD default.
+message NamespaceModeValue {
+	NamespaceMode value = 1;
+}
+```
+
+### Runtime Support
```

Comment:
How are the mapped UIDs applied to the container image layers? Does docker `chown` the files when the image is unpacked? What happens if a container running in the host user namespace shares a base layer with a container that is running in a different user namespace?

Reply:
Correct

Reply:
If you run with host user namespace, won't docker daemon pull and create completely different set of layers?

Reply:
I assumed @yujuhong was referring to the ability to override (for the `--privileged`/admin use case) with `docker run --userns=host` which starts a container using the same layers already downloaded and chown'ed as the shifted ID range. It is a bit kludgy but strongly requested to be possible for certain use cases. If you add a `--user` declaration to that run, you now have a non-shifted non-privileged user running the container process, and probably not having access to many of the files in the layers.

Reply:
Yes, @estesp that's what I meant. When per-pod configuration is supported in the future, this can cause great confusion for the users. I don't see a clear way to mitigate this...